keps/sig-api-machinery/20190621-apply-scalability.md

```diff
@@ -0,0 +1,154 @@
+The size of managedFields is similar to the size of the object. It increases etcd size, and cache size of the apiserver and clients.
+
+#### On update and apply: 
+
+Converting managedFields from the api representation into the internal representation causes many small object allocations.
+Some structured-merge-diff operations (like compare) also create many small objects that are not needed. Most “field set” operations are not done in place.
+
+### Experiments
+
+The following experiments convinced us that the problem is mostly the allocations of small objects. Latency in etcd because of read/write with bigger objects, and/or data transfer was not a significant issue.
+
+For all these scenarios, we’ve created a deployment with 5,000 pods. Pods that have the managedFields (when server-side apply is enabled) are obviously larger. Here are some numbers for the comparison, with the objects in protobuf format:
+
+- Server-side apply disabled:
+  - A pod is about 8,137 bytes
+  - 5,000 pods is 40,506,271 bytes
+- Server-side apply enabled:
+  - A pod is about 13,164 bytes (~62% larger)
+  - 5,000 pods is 67,182,690 bytes
+
+All the following scenarios have been tested by LISTing all the pods, then running 250 patches per second while still LISTing the pods at the same time.
+
+1. Server-side apply disabled, 
+2. Server-side apply enabled, 
+3. Server-side apply disabled, but with an extra annotation. The annotation is about 5,000 bytes to compensate for the difference in object size with server-side apply.
+4. Server-side apply disabled, with 320 small annotations. Those annotations total about 5,000 bytes to simulate the small object allocation.
+
+The average latency are as follows:
+
+| Scenario | LIST | PATCH | LIST while PATCH |
+|---|---|---|---|
+| 1. |360.4 ms | 8.0 ms | 361.0 ms |
+| 2. | 1294.5 ms | 38.5 ms | 2529.6 ms |
+| 3. | 480.8 ms | 11.3 ms | 500.2 ms |
+| 4. | 1260.1 ms | 9.7 ms | 1503.2 ms |
+
+### Goals
+
+TODO
+
+### Non-Goals
+
+TODO
+
+## Proposal
+
+We have considered a few overlapping solutions:
+
+### Only send managedFields if requested:
+
+Add a query parameter (or some other mechanism) to allow users to request to see the managedFields field in metadata. If the parameter is missing,  strip the field on reads and writes (in responses). The benefit here is we can skip serializing the field most of the time. Most clients won’t and shouldn’t have to understand how to read this field. So making them receive it is a waste of resources, and confusing.
+
+### Store managedFields as a blob
+
+Instead of a highly structured nested map, with lots of small fields, we could store it as a blob. The blob would save on allocation/GC of small objects when deserializing. We could also gzip the entire thing to save space. The downside is that it would make the field harder/impossible for a user to understand. This is mitigated by combining with solution 1, or by client-go/kubectl unzipping the field automatically.
```

Comment:
Alternatively revisit the serialization and choose a slightly more complicated serialization but with less size.  Maps are very expensive in memory, strings are less expensive, integers and slices are basically free.

For instance, an array with index references to protobuf fields, or internal index references.  Slices should be mostly free in deserialization, although field names will each require an allocation.

Reply:
Note that you could store it as a json blob fairly easily in etcd and it wouldn't be that difficult for clients that care.  You'd pay the cost to the deserializer to check the syntax (in json) but protobuf would be free.

Basically a typed RawExtension that was less generic (uses json.Message / equiv to avoid deserialization).  You'd still have to check for valid structure before saving to etcd, but you could avoid the check on read.

Reply:
We've done this in a couple spots in the openshift API.  As long as you have good helpers for digesting it, using a runtime.Unknown isn't a bad way to pass the bits around.

Reply:
1. Storing json blob helps only one way (i.e. read path) where we don't need to recompute it. On patches, we still need to recompute it, so deserialize, modify and serialize, so the overhead there will still be the same.

2. Forgive my ignorance (I didn't look into exact representation of managed fields), but I actually agree with Clayton`s suggestion that we should consider more complicated form of storing that data (possibly with good helpers with digestion).
Assuming that managed-fields is actually a nested map, we should definitely consider flattening it, e.g.:
["metadata.name": "name", "spec.nodeName": "nodeName", ...]
While maps are expensive, a single map is actually not that bad, given we already have maps in our objects.
If we would combine it with protobuf tags (e.g. "metadata.name" == "1.1") we could save a lot.

Reply:
> 1. Storing json blob helps only one way (i.e. read path) where we don't need to recompute it. On patches, we still need to recompute it, so deserialize, modify and serialize, so the overhead there will still be the same.

Yeah, I think we all agree on that :-/ That's why we also wanted to store it in a format that was convenient for patch, that is, in internal representation in protobuf format.

> Assuming that managed-fields is actually a nested map, we should definitely consider flattening it

We're storing the "set" of fields that are owned by each actor of the system. Today, we're using a trie, which I think is "complicated with less size". We could try using a list of such fields with their full path, which would be less complicated and bigger, and possibly faster.

Reply:
> We could try using a list of such fields with their full path, which would be less complicated and bigger, and possibly faster.

I'm pretty sure it will be faster. (we both agree it will be less complicated too).
And I actually would like to see data about the impact on size - I'm not sure it's that significant. And I think we should also consider what I mentioned about using protobuf tags instead of names, and that should significantly help (I actually kind of hope that it may be enough).

Reply:
> And I think we should also consider what I mentioned about using protobuf tags instead of names

Quick notes on that.

Pros:
- Shorter than strings, leading to smaller representation both in memory and serialized,
- Less allocations than strings,

Cons:
- API makes no sense unless you know what each protobuf number means, absolutely NOT human readable,
- Protobuf numbers are not in the API at all, there is no coupling between the two: I don't know how we would implement it, and certainly not how people would use it,
- No way to get this to work for CRs (I've heard rumors that we might get protobuf for CRs, but we don't have anything yet).

I'm fairly against it. I think it would really hurt the API for an uncertain performance boost.

> I'm pretty sure it will be faster. (we both agree it will be less complicated too).

Are you suggesting we do that on top of the "RawMessage", or we just serialize as lists of strings directly?
We would still have to parse these and generate the trie that we use internally to run the set operations that we have.

Reply:
Re cons:
1) I agree it's much less human-readable, but I don't actually agree it's "absolutely NOT human readable". It's possible for a human to get those numbers and parse it. And it's really easy to create helpers to those conversions automatically for humans.
2) I actually would say that these are kind of API. Auto-generated protobufs are part of the "api" repo and the tags are all there, e.g. for a Pod:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/core/v1/generated.proto#L2624
3) CRs are more tricky, but I actually can imagine doing first a simple check to see if protobuf exists for a given type and if not using names instead of tags.

So the cons are relatively weak in my opinion.

> Are you suggesting we do that on top of the "RawMessage", or we just serialize as lists of strings directly?
We would still have to parse these and generate the trie that we use internally to run the set operations that we have.

I wouldn't do RawMessage, but I'm also not against it. Simply list of pairs of strings, or maybe even map[string]string should make that much better.

> We would still have to parse these and generate the trie that we use internally to run the set operations that we have.

That makes me thing that slice of pairs of strings is better. If we would enforce that keys are sorted in trie order (which is what we can easily do and introduce validation), then building a trie from that representation (and vice versa) is trivial.

Reply:
I'm not sure why we need "pairs". We're simply tracking a list of fields, if we have a single string representing the path to a specific field, it's good enough.

Reply:
Hmm - we basically need mapping from field to its owner, right? That's why I say we need pairs.

Also to copy from our offline conversation about protos:
well - we need more numbers, but basically conceptually tags save us the biggest amount of data (in json vs proto) and there will be much less data to serialize/deserialize
but if you show me data of experiments that shows that this is not needed, I would be more than happy with it (I just think that simple flattening may not be enough)

Reply:
I'm biased towards trying to be embed something recognizeable to most clients and I think JSON meets that need better than proto.  If we have many, small, nested maps perhaps it makes more sense to keep a hierarchical structure and simply use short slices instead.  The gains in conversion being able to do a simple pointer assignment may offset the o(1) behavior of maps if your slices are short.

Reply:
As I mentioned above, I completely buy the argument of human-readability, if we can get enough gains from json-like format.

That said, I don't really understand why you're pushing for NOT flattening that - flattening it would visibly reduce amount of allocations, right? And it's similarly understandable for users.

Reply:
Let's take this thread to https://github.com/kubernetes-sigs/structured-merge-diff/issues/99.

Proto numbers would make the objects non-self describing and don't currently work for CRDs.